std/experimental/typecons.d

```diff
@@ -0,0 +1,731 @@
+ * a copy, it is not possible to affect the original with a wrapped structure.
+ */
+template wrap(Targets...)
+if (Targets.length >= 1 && allSatisfy!(isInterface, Targets))
+{
+    import std.meta : staticMap;
+
+    static if (!allSatisfy!(isMutable, Targets)) {
+        alias wrap = .wrap!(staticMap!(Unqual, Targets));
+    } else {
+        // strict upcast
+        auto wrap(Source)(inout Source src) @trusted pure nothrow
+        if (Targets.length == 1 && is(Source : Targets[0]))
+        {
+            alias T = Select!(is(Source == shared), shared Targets[0], Targets[0]);
+            return dynamicCast!(inout T)(src);
+        }
+        // structural upcast
+        template wrap(Source)
+        if (!allSatisfy!(Bind!(isImplicitlyConvertible, Source), Targets))
+        {
+            auto wrap(inout Source src)
+            {
+                static assert(implementsInterface!(Source, Targets),
+                              "Source "~Source.stringof~
+                              " does not have structural conformance to "~
+                              Targets.stringof);
+
+                alias T = Select!(is(Source == shared), shared Impl, Impl);
+                return new inout T(src);
+            }
+
+            // list of FuncInfo
+            alias TargetMembers = UniqMembers!(ConcatInterfaceMembers!(Targets));
+            // list of function symbols
+            alias SourceMembers = GetOverloadedMethods!Source;
+
+            static if (is(Source == class) || is(Source == interface))
+                alias StructuralType = Object;
+            else static if (is(Source == struct))
+                alias StructuralType = Source;
+
+            // Check whether all of SourceMembers satisfy covariance target in TargetMembers
+            // Internal wrapper class
+            final class Impl : Structural!StructuralType, Targets
+            {
+            private:
+                Source _wrap_source;
+
+                this(       inout Source s)        inout @safe pure nothrow { _wrap_source = s; }
+                this(shared inout Source s) shared inout @safe pure nothrow { _wrap_source = s; }
+
+                static if (is(Source == class) || is(Source == interface))
+                {
+                    // BUG: making private should work with NVI.
+                    protected final inout(Object) _wrap_getSource() inout @safe
+                    {
+                        return dynamicCast!(inout Object)(_wrap_source);
+                    }
+                }
+                else
+                {
+                    // BUG: making private should work with NVI.
+                    protected final inout(Source) _wrap_getSource() inout @safe
+                    {
+                        return _wrap_source;
+                    }
+                }
+
+                import std.conv : to;
+                import std.functional : forward;
+                template generateFun(size_t i)
```

Comment:
@MartinNowak, that may be true, but that just means 'wrap' is unable to support those types. I did introduce a wrapperSignature in this commit: https://github.com/JesseKPhillips/phobos/commit/29ecc4afd4bd4d6e26d76e27a7ff65e69d707e1d